landlab/components/lake_fill/lake_fill_barnes.py

```diff
@@ -0,0 +1,2149 @@
+        ...           'instantiation.')
+        ValueError was raised: Pit is overfilled due to creation of two outlets as the minimum gradient gets applied. Suppress this Error with the ignore_overfill flag at component instantiation.
+
+        Suppress this behaviour with ignore_overfill:
+
+        >>> z[:] = z_init
+        >>> lmb = LakeMapperBarnes(mg, method='steepest',
+        ...                        fill_flat=False, track_lakes=True,
+        ...                        ignore_overfill=True)
+        >>> lmb.run_one_step()
+        >>> lmb.lake_dict == {8: deque([7]), 16: deque([15, 9, 14, 22])}
+        True
+        >>> lmb.run_one_step()
+        >>> np.allclose(lmb.lake_areas, np.array([ 16.,  4.]))  # found them!
+        True
+
+        The component can redirect flow to account for the fills that have
+        been carried out (all necessary fields get updated):
+
+        >>> z[:] = z_init
+        >>> fd.run_one_step()
+        >>> init_flowdirs = mg.at_node['flow__receiver_node'].copy()
+        >>> fa.run_one_step()
+        >>> init_areas = mg.at_node['drainage_area'].copy()
+        >>> init_qw = mg.at_node['surface_water__discharge'].copy()
+
+        >>> lmb = LakeMapperBarnes(mg, method='steepest',
+        ...                        fill_flat=False, track_lakes=True,
+        ...                        redirect_flow_steepest_descent=False,
+        ...                        ignore_overfill=True)
+        >>> lmb.run_one_step()
+        >>> np.all(mg.at_node['flow__receiver_node'] == init_flowdirs)
+        True
+
+        >>> lmb = LakeMapperBarnes(mg, method='steepest',
+        ...                        fill_flat=False, track_lakes=True,
+        ...                        redirect_flow_steepest_descent=True,
+        ...                        ignore_overfill=True)
+        >>> lmb.run_one_step()
+        >>> np.all(mg.at_node['flow__receiver_node'] == init_flowdirs)
+        False
+
+        However, note that unless the reaccumulate_flow argument is also
+        set, the 'drainage_area' and 'surface_water__discharge' fields
+        *won't* also get updated:
+
+        >>> np.all(mg.at_node['drainage_area'] == init_areas)
+        True
+        >>> np.all(mg.at_node['surface_water__discharge'] == init_qw)
+        True
+
+        >>> lmb = LakeMapperBarnes(mg, method='steepest',
+        ...                        fill_flat=False, track_lakes=True,
+        ...                        redirect_flow_steepest_descent=True,
+        ...                        reaccumulate_flow=True,
+        ...                        ignore_overfill=True)
+        >>> lmb.run_one_step()
+        >>> np.all(mg.at_node['drainage_area'] == init_areas)
+        False
+        >>> np.all(mg.at_node['surface_water__discharge'] == init_qw)
+        False
+
+        Be sure to set both redirect_flow_steepest_descent and
+        reaccumulate_flow to True if you want to reaccum
```

Comment:
Alrighty. I've made it to the end of this document. 

One major comment: You have a lot of assertion errors that you describe and test in these docstrings. I think it would be MUCH better if you raised the appropriate type of error with a good error message _within_ the code. Right now the only way the user gets information about what the different assertion methods means is because of the text you've printed within the docstring try-except blocks. If the message is within the error itself, then a user has a hope of debugging without pouring through the docstrings. 

Along with this, I'd move the tests of these sorts of assertions/errors into unit tests. 

I forget if you were away while this change got made, but we are no longer using nose. So instead of nose `assert_raises` do the following:

```python
import pytest

with pytest.raises(ValueError):
    #do the thing that causes the value error
```

Reply:
Now raising Errors wherever these are actually likely to get seen. A few asserts remain in places that could become problems down the line if this ever gets modified to e.g. route-to-many. Some tests remain inline, as I think they're helpful. Others unittested.